tests/testUtils.py

```diff
@@ -50,6 +51,65 @@ class BlockLogAction(EnumType):
 addEnum(BlockLogAction, "prune_transactions")
 
 ###########################################################################################
+
+class Timeout:
+
+    def __init__(self, value, sleep_interval=None, leeway=None):
+        self.toCalculate = True if value == -1 else False
+        if value is not None:
+            assert isinstance(value, (int))
+            assert value >= -1
+        if sleep_interval is not None:
+            assert isinstance(sleep_interval, (int))
+            assert sleep_interval > 0
+        self.value = value
+        self.sleep_interval = sleep_interval
+        self.leeway = leeway
+
+    def __str__(self):
+        append = "[calculated based on block production]" if self.toCalculate else ""
+        desc = None
+        if self.value is None:
+            desc = "defaulted"
+        elif self.value >= 0:
+            desc = "%d sec" % (self.value)
+        else:
+            desc = ""
+        return "Timeout %s%s" % (desc, append)
+
+    def convert(self, startBlockNum, endBlockNum):
+        if self.value is None or self.value != -1:
+            return
+
+        timeout = self.leeway if self.leeway is not None else 10
+        if (endBlockNum > startBlockNum):
+            # calculation is performing worst case (irreversible block progression) which at worst will waste 5 seconds
+            blocksPerWindow = 12
+            blockWindowsToWait = (endBlockNum - startBlockNum + blocksPerWindow - 1) / blocksPerWindow
+            secondsPerWindow = blocksPerWindow / 2
+            timeout += blockWindowsToWait * secondsPerWindow
+
+        self.value = timeout
+        if self.sleep_interval is None:
+            self.sleep_interval = max(timeout / 10, 1)
+
+    def sleepTime(self, default=None):
+        time = self.sleep_interval if self.sleep_interval is not None else default
+        return time if time is not None else Timeout.default_interval_seconds
+
+    def asSeconds(self):
+        assert self.value != -1, "Timeout.calculate was used, but calculate method never called."
```

Comment:
There is no `calculate` method.  Perhaps this should read: "Timeout automatic calculation was used but convert method was never called."

Reply:
fixed

Reply:
This still says `but calculate method never called`.

I'm not sure how useful saying `Timeout.calculate was used` is, either.  I misinterpreted it the first time, not realizing it refers to the attribute set below.  It may make sense once it stops talking about a nonexistent `calculate` method.

Reply:
Yeah, fixed it to say convert and added more information.  This is intended to tell the test writer that either somewhere in the chain of methods it needs to call convert or else they incorrectly used WaitSpec.calculate(...)